contracts/test/LightClient.t.sol

```diff
@@ -0,0 +1,450 @@
+// SPDX-License-Identifier: Unlicensed
+
+/* solhint-disable contract-name-camelcase, func-name-mixedcase, one-contract-per-file */
+
+pragma solidity ^0.8.0;
+
+// Libraries
+import "forge-std/Test.sol";
+import { IPlonkVerifier as V } from "../src/interfaces/IPlonkVerifier.sol";
+
+// Target contract
+import { LightClient as LC } from "../src/LightClient.sol";
+import { LightClientTest as LCTest } from "./mocks/LightClientTest.sol";
+import { BN254 } from "bn254/BN254.sol";
+
+/// @dev Common helpers for LightClient tests
+contract LightClientCommonTest is Test {
+    LCTest lc;
+    uint32 public constant BLOCKS_PER_EPOCH_TEST = 3;
+    LC.LightClientState public genesis;
+    // this constant should be consistent with `hotshot_contract::light_client.rs`
+    uint64 internal constant STAKE_TABLE_CAPACITY = 10;
+
+    /// @dev initialized ledger like genesis and system params
+    function init() public {
+        string[] memory cmds = new string[](4);
+        cmds[0] = "diff-test";
+        cmds[1] = "mock-genesis";
+        cmds[2] = vm.toString(BLOCKS_PER_EPOCH_TEST);
+        cmds[3] = vm.toString(STAKE_TABLE_CAPACITY / 2);
+
+        bytes memory result = vm.ffi(cmds);
+        (LC.LightClientState memory state, bytes32 votingSTComm, bytes32 frozenSTComm) =
+            abi.decode(result, (LC.LightClientState, bytes32, bytes32));
+
+        genesis = state;
+        lc = new LCTest(genesis, BLOCKS_PER_EPOCH_TEST);
+        bytes32 expectedStakeTableComm = lc.computeStakeTableComm(state);
+        assertEq(votingSTComm, expectedStakeTableComm);
+        assertEq(frozenSTComm, expectedStakeTableComm);
+    }
+
+    function assertEq(BN254.ScalarField a, BN254.ScalarField b) public {
+        assertEq(BN254.ScalarField.unwrap(a), BN254.ScalarField.unwrap(b));
+    }
+
+    /// @dev assertEq for `struct LightClientState`
+    function assertEqState(LC.LightClientState memory a, LC.LightClientState memory b) public {
+        assertEq(a.viewNum, b.viewNum);
+        assertEq(a.blockHeight, b.blockHeight);
+        assertEq(a.blockCommRoot, b.blockCommRoot);
+        assertEq(a.feeLedgerComm, b.feeLedgerComm);
+        assertEq(a.stakeTableBlsKeyComm, b.stakeTableBlsKeyComm);
+        assertEq(a.stakeTableSchnorrKeyComm, b.stakeTableSchnorrKeyComm);
+        assertEq(a.stakeTableAmountComm, b.stakeTableAmountComm);
+        assertEq(a.threshold, b.threshold);
+    }
```

Comment:
This can be simpified as we can use `abi.encode` to compare structs.
```suggestion
     /// @dev assertEq for `struct LightClientState`
    function assertEqState(LC.LightClientState memory a, LC.LightClientState memory b) public {
        assertEq(abi.encode(a), abi.encode(b));
    }
```

Reply:
ah that's a neat idea!